core/macros/src/class.rs

```diff
@@ -0,0 +1,703 @@
+                    Ok(Default::default())
+                }
+            },
+            |c| c.body.clone(),
+        );
+
+        quote! {
+            impl boa_engine::class::Class for #class_ty {
+                const NAME: &'static str = #class_name;
+                const LENGTH: usize = #arg_count;
+
+                fn data_constructor(
+                    this: &boa_engine::JsValue,
+                    args: &[boa_engine::JsValue],
+                    context: &mut boa_engine::Context
+                ) -> boa_engine::JsResult<Self> {
+                    #constructor_body
+                }
+
+                fn init(builder: &mut boa_engine::class::ClassBuilder) -> boa_engine::JsResult<()> {
+                    // Add all statics.
+                    #(#builder_statics)*
+
+                    // Add all accessors.
+                    #(#accessors)*
+
+                    // Add all methods to the class.
+                    #(#builder_methods)*
+
+                    Ok(())
+                }
+            }
+        }
+    }
+}
+
+impl VisitMut for ClassVisitor {
+    // Allow similar names as there are no better ways to name `getter` and `setter`.
+    #[allow(clippy::similar_names)]
+    fn visit_impl_item_fn_mut(&mut self, item: &mut ImplItemFn) {
+        // If there's a `boa` argument, parse it.
+        let has_ctor_attr = take_path_attr(&mut item.attrs, "constructor");
+        let has_getter_attr = take_path_attr(&mut item.attrs, "getter");
+        let has_setter_attr = take_path_attr(&mut item.attrs, "setter");
+        let has_method_attr = take_path_attr(&mut item.attrs, "method");
+
+        if has_getter_attr {
+            self.getter(item);
+        }
+
+        if has_setter_attr {
+            self.setter(item);
+        }
+
+        if has_ctor_attr {
+            self.constructor(item);
+        }
+
+        // A function is a method if it has a `#[boa(method)]` attribute or has no
+        // method-type related attributes.
+        if has_method_attr || !(has_getter_attr || has_ctor_attr || has_setter_attr) {
+            self.method(&item.sig.span(), item);
+        }
+
+        syn::visit_mut::visit_impl_item_fn_mut(self, item);
+    }
+}
+
+#[derive(Debug)]
+struct ClassArguments {
+    name: Option<String>,
+}
+
+impl Parse for ClassArguments {
+    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
+        let args: Punctuated<Meta, Token![,]> = Punctuated::parse_terminated(input)?;
+        let mut name = None;
+
+        for arg in &args {
+            match arg {
+                Meta::NameValue(MetaNameValue {
+                    path,
+                    value: Expr::Lit(lit),
+                    ..
+                }) if path.is_ident("name") => {
+                    name = Some(match &lit.lit {
+                        Lit::Str(s) => Ok(s.value()),
+                        _ => Err(syn::Error::new(lit.span(), "Expected a string literal")),
+                    }?);
```

Comment:
This doesn't handle the case when name is defined twice, right?

Reply:
There are other failsafes for this in basic cases, but no it doesn't prevent you from going out of your way. This would be similar to setting the prototype's key `name` twice, which is also allowed.